File: Scala/PageRank/src/main/scala/pageRank.scala
import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.HashPartitioner
import org.apache.spark.Partitioner

class DomainPartitioner(numParts: Int) extends Partitioner{
	override def numPartitions: Int = numParts
	override def getPartition(key: Any): Int={
		val partId = (key.toString.toInt/numParts)
		println("NodeId: "+key+ "partId:" +partId)
		if (partId>=numParts){
			numParts - 1
		}
		else{
			partId
		}
	}
	override def equals(other: Any): Boolean = other match{
		case dnp: DomainPartitioner=>
			dnp.numPartitions==numPartitions
		case _=>
			false
	}
}

object pageRank{
	def main(args: Array[String]){
		val conf = new SparkConf().setMaster("local").setAppName("word count")
                val sc   = new SparkContext(conf)
		
		val inputFile = "file:"+args(0)
		val readInput = sc.textFile(inputFile)		
	 	
		val blinks = readInput.map(x=>x.split(" ")).map(x=>(x(0),x(1)))
		val links = blinks.groupByKey().partitionBy(new DomainPartitioner(4)).persist()
		
		var ranks = links.mapValues(x=>1.0)
		
		for(i<-0 until args(1).toInt){
			val contrib = links.join(ranks).flatMap{
				case(nodeId,(dsts,rank))=>dsts.map(dst=>(dst,rank/dsts.size))
			}
			ranks = contrib.reduceByKey((x,y)=>x+y).mapValues(x=>0.15+0.85*x)

		}
		ranks.saveAsTextFile("file:"+args(2))
	}
}

File: Python/pageRank/pageRank.py
from pyspark.conf import SparkConf
from pyspark.context import SparkContext
import sys
def formatInput(x):
	values = x.split(' ')
	return (int(values[0]),int(values[1]))
def calContrib(dstList, rank):
	length = len(dstList)
	for j in dstList:
		yield (j,rank/length)

def partFunc(pageId,numParts):
	partNum = pageId/numParts
	print("Node Id: %i, partNum: %i"%(pageId,partNum))
	return partNum

if __name__ == "__main__":
	#run format <pageRank, inputFile, iterations, outputFile, numMachines>
	if len(sys.argv) != 5:
		print("Please verify input format")
		sys.exit(-1)
	conf = SparkConf().setMaster("local").setAppName("pageRank")
	sc   = SparkContext(conf=conf)
	
	inputFile = 'file:'+sys.argv[1]
	readInput = sc.textFile(inputFile)
	#format the input in (src,list(dst)) format
	links = readInput.map(lambda x:formatInput(x))
	links = links.groupByKey()
	#Partition the graph using range partitioner
	numMachines = int(sys.argv[4])
	links = links.partitionBy(numMachines, lambda x: partFunc(x,numMachines)).persist()
	
	
	#initialize ranks by (src,1)
	ranks = links.mapValues(lambda x: 1.0)
	
	for i in range(0,int(sys.argv[2])):
		#join will create a list of format (src,(list(dst),rank_src)) format
		#use flatMap to get it in (dst,rank_src) format
		contrib = links.join(ranks).flatMap(lambda x: calContrib(x[1][0],x[1][1]))
		#now use reduceByKey to get contribution for the node and then calculate pageRank by 0.15+0.85*contrib(node)
		ranks   = contrib.reduceByKey(lambda x,y: x+y).mapValues(lambda x: 0.15+0.85*x)

	#now save the result to outputFolder
	ranks.saveAsTextFile('file:'+sys.argv[3])
	
